sources/kms/grant.go

```diff
@@ -0,0 +1,220 @@
+package kms
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/service/kms"
+	"github.com/overmindtech/aws-source/sources"
+	"github.com/overmindtech/sdp-go"
+)
+
+func grantOutputMapper(_ context.Context, _ *kms.Client, scope string, _ *kms.ListGrantsInput, output *kms.ListGrantsOutput) ([]*sdp.Item, error) {
+	items := make([]*sdp.Item, 0)
+
+	for _, grant := range output.Grants {
+		attributes, err := sources.ToAttributesCase(grant, "tags")
+		if err != nil {
+			return nil, err
+		}
+
+		// This should never happen.
+		if grant.GrantId == nil {
+			return nil, &sdp.QueryError{
+				ErrorType:   sdp.QueryError_OTHER,
+				ErrorString: "grantId is nil",
+			}
+		}
+
+		// The uniqueAttributeValue for this is the combination of the keyID and grantId
+		// i.e., "cf68415c-f4ae-48f2-87a7-3b52ce/grant-id"
+		err = attributes.Set("uniqueName", fmt.Sprintf("%s/%s", *grant.KeyId, *grant.GrantId))
+		if err != nil {
+			return nil, err
+		}
+
+		item := sdp.Item{
+			Type:            "kms-grant",
+			UniqueAttribute: "uniqueName",
+			Attributes:      attributes,
+			Scope:           scope,
+		}
+
+		if grant.KeyId != nil {
+			arn, errA := sources.ParseARN(*grant.KeyId)
+			if errA != nil {
+				return nil, &sdp.QueryError{
+					ErrorType:   sdp.QueryError_OTHER,
+					ErrorString: fmt.Sprintf("failed to parse keyID: %s", *grant.KeyId),
+				}
+			}
+
+			// +overmind:link kms-key
+			item.LinkedItemQueries = append(item.LinkedItemQueries, &sdp.LinkedItemQuery{
+				Query: &sdp.Query{
+					Type:   "kms-key",
+					Method: sdp.QueryMethod_GET,
+					Query:  strings.TrimPrefix(arn.Resource, "key/"),
+					Scope:  scope,
+				},
+				BlastPropagation: &sdp.BlastPropagation{
+					// These are tightly linked
+					// Adding or revoking/retiring a grant can allow or deny permission to the KMS key for the grantee.
+					In:  true,
+					Out: true,
+				},
+			})
+		}
+
+		var principals []string
+		if grant.GranteePrincipal != nil {
+			principals = append(principals, *grant.GranteePrincipal)
+		}
+
+		if grant.RetiringPrincipal != nil {
+			principals = append(principals, *grant.RetiringPrincipal)
+		}
+
+		// Valid principals include
+		// - Amazon Web Services accounts
+		// - IAM users,
+		// - IAM roles,
+		// - federated users,
+		// - assumed role users.
+		// principals: https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_identifiers.html#identifiers-arns
+		/*
+			arn:aws:iam::account:root
+			arn:aws:iam::account:user/user-name-with-path
+			arn:aws:iam::account:role/role-name-with-path
+			arn:aws:sts::account:federated-user/user-name
+			arn:aws:sts::account:assumed-role/role-name/role-session-name
+			arn:aws:sts::account:self
+
+			dynamodb.us-west-2.amazonaws.com
+
+			The following are not supported
```

Comment:
I couldn't find a resource that would be used to link these type of principals.
Maybe we should create a ticket for that.